doc/development/guide/architecture.rst

```diff
@@ -0,0 +1,204 @@
+the library. It includes basic functionality that is shared for quantum
+devices, independent of the quantum information model used. PennyLane gives
+access to multiple simulators and hardware chips through its plugins; each of
+these devices is implemented as a custom class. These classes have the
+``Device`` class as their parent class.
+
+The purpose of the ``Device`` class can be summarized as:
+
+* Providing a common API to execute a quantum circuit and request
+  the measurement of the associated observable.
+* Providing an easy way of developing a new device for PennyLane
+
+Qubit based devices can use shared utilities by using the
+:class:`~.QubitDevice`.
+
+QNodes
+******
+
+A  quantum node or ``QNode`` (represented by a subclass to
+:class:`~.BaseQNode`) is an encapsulation of a function
+:math:`f(x;\theta)=R^m\rightarrow R^n` that is executed using quantum
+information processing on a quantum device.
+
+Apart from incorporating quantum functions, ``QNodes`` also offer custom
+quantum differentiation rules. Using the so-called `parameter-shift rules
+<https://pennylane.ai/qml/glossary/parameter_shift.html>`__, many quantum
+functions can be expressed through the linear combination of other quantum
+functions. As these rules allow quantum gradients to be obtained from
+``QNodes``, hybrid computations may include ``QNodes`` as part of training deep
+learnings models.
+
+PennyLane offers the following qnode types and differentiation rules:
+
+* :class:`~.QubitQNode`: qubit parameter-shift rule
+* :class:`~.CVQNode`: CV parameter-shift rule
+* :class:`~.JacobianQNode`: finite differences
+* :class:`~.DeviceJacobianQNode`: device based gradient rule
+* :class:`~.PassthruQNode`: classical backpropagation
+* :class:`~.ReversibleQNode`: reversible backpropagation
+
+These qnode types are available to users through the :func:`~.qnode` decorator by
+passing the user-facing ``diff_method`` option. This decorator then uses the
+:func:`~.QNode` constructor function to create the specific type of qnode based on
+the device, interface, and quantum function.
+
+A widespread representation of quantum circuits is by creating a `Directed
+Acyclic Graph (DAG)
+<https://pennylane.ai/qml/glossary/hybrid_computation.html#directed-acyclic-graphs>`__
+and representing quantum operations within such a graph. Each ``QNode``
+represents the quantum circuit by building such a DAG by creating a
+:class:`~.CircuitGraph` instance.
+
+For further details on QNodes, and a full list of QNodes, refer to the
+:doc:`/code/qml_qnodes` module.
+
+Interfaces
+**********
+
+The integration between classical and quantum computations is encompassed by
+interfaces. Differentiable and Jacobian ``QNodes`` have interfacing functions
+that provide a 'wrapper' around QNodes such. These wrappers further transform
+the ``QNode`` such that the quantum gradient rules of the QNodes are registered
+to the machine learning interface via a custom gradient class or function.
```

Comment:
I suppose we could go into further detail here with what exactly an interface function does.

* It wraps the QNode, returning a QNode that accepts and returns the core data structure of the classical machine learning library (e.g., a TF tensor, Torch tensor, Autograd NumPy array, etc).

* It _unwraps_ the input data structures to simple NumPy arrays, so that the quantum device can execute the
  users quantum function.

* It registers the `QNode.jacobian()` method as a custom gradient method, so that the machine learning library can 'backpropagate' across the QNode, when integrated into a classical computation.

@co9olguy has a nice diagram that represents the unwrapping and evaluation performed by interfaces :)